Framework/API/src/MatrixWorkspace.cpp

```diff
@@ -83,6 +84,31 @@ MatrixWorkspace::~MatrixWorkspace() {
   }
 }
 
+void MatrixWorkspace::setIndexTranslator(
+    Indexing::SpectrumNumbers &&spectrumNumbers,
+    Indexing::DetectorIDs &&detectorIDs) {
+  setIndexTranslator(Indexing::IndexTranslator(std::move(spectrumNumbers),
+                                               std::move(detectorIDs)));
+}
+
+void MatrixWorkspace::setIndexTranslator(
+    Indexing::IndexTranslator &&translator) {
+  if (translator.size() != getNumberHistograms())
+    throw std::runtime_error("MatrixWorkspace::setIndexTranslator: Translator "
+                             "size does not match number of histograms in "
+                             "workspace");
+
+  auto &spectrumNumbers = translator.spectrumNumbers();
+  auto &detectorIDs = translator.detectorIDs();
+  for (size_t i = 0; i < getNumberHistograms(); ++i) {
+    auto &spectrum = getSpectrum(i);
+    spectrum.setSpectrumNo(spectrumNumbers[i]);
+    if (!detectorIDs.empty())
```

Comment:
@SimonHeybrock. I'm not sure about this option. I quite liked the idea about having a single `IndexTranslator` that could manage all self-consistency internally. 

Somehow I think that what you really want to do from the algorithm perspective is:

1. Take the existing `IndexTranslator` on the `InputWorkspace`. Possibly by cloning the input workspace to make your `OutputWorkspace`.
2. Overwrite just the spectrum numbers on the `IndexTranslator` on the `OutputWorkspace`. This should invoke COW. but we would keep the `DetectorID` parts.

This supposes that the `IndexTranslator` is carried by the `MatrixWorkspace`. So this is something that you've already mentioned you'd only do in a later refactoring step. It then requires (unless you use a `smart_ptr`) that you are going to include `IndexTranslator` in `MatrixWorkspace.h`, so (as an aside) with your previous question you may as well have:
```
template <class... T>
void setIndexTranslator(T &&... args);
```